plugins/chat/spec/services/chat/auto_join_channels_spec.rb

```diff
@@ -0,0 +1,359 @@
+
+          Fabricate(:user, trust_level:, last_seen_at:)
+          Fabricate(:admin, trust_level:, last_seen_at:)
+
+          expect { job }.to change { Chat::UserChatChannelMembership.count }.from(0).to(1)
+        end
+
+        it "limits the number of users who can be auto-joined to SiteSetting.max_chat_auto_joined_users" do
+          SiteSetting.max_chat_auto_joined_users = 1
+
+          _user_1 = Fabricate(:user, trust_level:, last_seen_at: 10.days.ago)
+          user_2 = Fabricate(:user, trust_level:, last_seen_at: 5.days.ago)
+
+          expect { job }.to change { Chat::UserChatChannelMembership.count }.from(0).to(1)
+          expect(Chat::UserChatChannelMembership.last.user).to eq(user_2)
+        end
+
+        it "doesn't automatically join users on deleted channels" do
+          auto_joinable_public_channel.update!(deleted_at: 1.day.ago)
+
+          Fabricate(:user, trust_level:, last_seen_at:)
+
+          expect { job }.not_to change { Chat::UserChatChannelMembership.count }.from(0)
+        end
+
+        it "doesn't automatically join users to channels that have reached the maximum user count" do
+          auto_joinable_public_channel.update!(user_count: SiteSetting.max_chat_auto_joined_users)
+
+          Fabricate(:user, trust_level:, last_seen_at:)
+
+          expect { job }.not_to change { Chat::UserChatChannelMembership.count }.from(0)
+        end
+
+        it "supports filtering down to a specific channel" do
+          Fabricate(:chat_channel, chatable: public_category, auto_join_users: true)
+
+          Fabricate(:user, trust_level:, last_seen_at:)
+
+          expect { described_class.call(channel_id: auto_joinable_public_channel.id) }.to change {
+            Chat::UserChatChannelMembership.count
+          }.from(0).to(1)
+        end
+
+        it "supports filtering down to a specific public category" do
+          Fabricate(:chat_channel, chatable: Fabricate(:category), auto_join_users: true)
+
+          Fabricate(:user, trust_level:, last_seen_at:)
+
+          expect { described_class.call(category_id: public_category.id) }.to change {
+            Chat::UserChatChannelMembership.count
+          }.from(0).to(1)
+        end
+
+        it "doesn't create duplicate memberships" do
+          user = Fabricate(:user, trust_level:, last_seen_at:)
+
+          Chat::UserChatChannelMembership.create!(user:, chat_channel: auto_joinable_public_channel)
+
+          expect { job }.not_to change { Chat::UserChatChannelMembership.count }.from(1)
+        end
+
+        it "automatically follows the channel in automatic mode" do
+          user = Fabricate(:user, trust_level:, last_seen_at:)
+
+          expect { job }.to change {
+            Chat::UserChatChannelMembership
+              .where(user:, chat_channel: auto_joinable_public_channel)
+              .where(following: true, join_mode: :automatic)
+              .count
+          }.from(0).to(1)
+        end
+
+        it "recalculates user count" do
```

Comment:
To pick up on my previous comment, here for example, this example is something happening when everything’s fine, but it feels a bit lost in all the other examples. Using contexts to describe various states will probably help to then group examples from the normal/happy path together and thus help to better understand what’s happening.

Reply:
Ah, I think I wasn’t clear, sorry :sweat_smile: I’m not suggesting to regroup everything into _one_ example, but to group the examples (having the various examples regarding the happy path, without any condition) next to each other.

Reply:
Gotcha 👍